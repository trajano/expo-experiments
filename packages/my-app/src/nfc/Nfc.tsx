/**
 * A context that provides NFC functionality. This starts up the NfcManager and provides a getTag function
 */
import {
  ComponentType,
  createContext,
  FC,
  PropsWithChildren,
  useCallback,
  useContext,
  useEffect,
  useMemo,
} from 'react';
import NfcManager, { NfcTech, TagEvent } from 'react-native-nfc-manager';

/**
 * Interface representing an empty context value for the Nfc context.
 * This interface can be expanded to include any values needed by the context.
 */
export interface Nfc {
  nfcManager: typeof NfcManager;
  getTagAsync: (...tech: NfcTech[]) => Promise<TagEvent | null>;
}

/**
 * The NfcContext provides an empty context value. This is exported
 * so that it can be used by tests, but generally the {@link NfcProvider}
 * is used.
 */
export const NfcContext = createContext<Nfc>({
  nfcManager: NfcManager,
  getTagAsync: () => Promise.reject(new Error('uninitialized context')),
});

/**
 * Props type for the NfcProvider component, which wraps its children in the NfcContext.
 */
export type NfcProps = PropsWithChildren;

/**
 * A provider component for the NfcContext.
 *
 * @param children The child components that will have access to the context.
 * @returns A provider that passes an empty value to all of its children.
 */
export const NfcProvider: FC<NfcProps> = ({ children }) => {
  const getTagAsync = useCallback(async (...nfcTechs: NfcTech[]) => {
    try {
      await NfcManager.requestTechnology(nfcTechs);
      return NfcManager.getTag();
    } finally {
      await NfcManager.cancelTechnologyRequest();
    }
  }, []);
  const value = useMemo<Nfc>(
    () => ({
      nfcManager: NfcManager,
      getTagAsync,
    }),
    [getTagAsync],
  );

  useEffect(() => {
    (async () => {
      await NfcManager.start();
    })();
  }, []);
  return <NfcContext.Provider value={value}>{children}</NfcContext.Provider>;
};

/**
 * Hook to consume the NfcContext. This hook allows components to access the empty context value.
 *
 * @returns The current context value, which is an empty object.
 */
export const useNfc = () => useContext(NfcContext);

/**
 * A higher-order component (HOC) that wraps a given component with the NfcProvider,
 * allowing the wrapped component to access the NfcContext.
 *
 * @param Component The component to be wrapped.
 * @returns A new component wrapped with the NfcProvider.
 */
export const WithNfc = <P extends object>(
  Component: ComponentType<P>,
): FC<P & NfcProps> => {
  const WrappedComponent = ({ ...props }: P & NfcProps) => (
    <NfcProvider>
      <Component {...(props as P)} />
    </NfcProvider>
  );

  WrappedComponent.displayName = `WithNfc(${Component.displayName ?? Component.name ?? 'Component'})`;

  return WrappedComponent;
};
