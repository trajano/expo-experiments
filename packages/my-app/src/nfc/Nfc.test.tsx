import { render, renderHook, screen } from '@testing-library/react-native';
import { Text } from 'react-native';
import { NfcContext, NfcProvider, useNfc, WithNfc } from './Nfc';
import { FC, PropsWithChildren } from 'react';

const TestComponent: FC = () => {
  const NfcProps = useNfc();

  return <Text testID="props">{JSON.stringify(NfcProps)}</Text>;
};

describe('Nfc', () => {
  it('Use provider', async () => {
    render(
      <NfcProvider>
        <TestComponent />
      </NfcProvider>,
    );
    expect(screen.getByTestId('props').props.children).toStrictEqual(
      JSON.stringify({ nfcManager: {} }),
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('Use Hook', async () => {
    let value = {
      nfcManager: jest.fn() as any,
      getTagAsync: jest.fn(),
    };
    const wrapper = ({ children }: PropsWithChildren) => {
      return (
        <NfcContext.Provider value={value}>{children}</NfcContext.Provider>
      );
    };
    const { result } = renderHook(() => useNfc(), { wrapper });
    expect(result.current).toEqual(value);
  });

  it('Use HoC', async () => {
    const TestedComponent = WithNfc(TestComponent);
    render(<TestedComponent />);
    expect(screen.getByTestId('props').props.children).toStrictEqual(
      JSON.stringify({ nfcManager: {} }),
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
