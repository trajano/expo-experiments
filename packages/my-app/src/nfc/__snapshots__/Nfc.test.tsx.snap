// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nfc Use HoC 1`] = `
<Text
  testID="props"
>
  {"nfcManager":{}}
</Text>
`;

exports[`Nfc Use provider 1`] = `
<Text
  testID="props"
>
  {"nfcManager":{}}
</Text>
`;
